### PR TITLE
New Internal Aiger Format

### PR DESCRIPTION
**Pull request by mvcisback**
The primary benefits are how much simpler composition and other bookkeeping become.
Some design choices:
- Focus on immutable types only. These come with off the shelf hashing as well as avoid common bugs due to state.
- Avoid encoding redudant information. I removed the header and other things from the representation. Many things will likely return as properties.

### CODE REVIEW

File: aiger/aig.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+    def __or__(self, other):
+        return par_compose(self, other)
+
+    @property
+    def _eval_order(self):
+        return list(toposort(_dependency_graph(self.gates)))
+    
+    def __call__(self, inputs, latches=None):
+        # TODO: Implement partial evaluation.
+        if latches is not None:
```
Comment (MarkusRabe) [inline aiger/aig.py]: latches is None

File: aiger/aig.py [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+    )
+
+
+def _const(val):
+    return Inverter(Ground()) if val else Ground()
+
+
+def source(outputs):
+    return AIG(
+        inputs=tuple(),
```
Comment (MarkusRabe) [inline aiger/aig.py]: frozenset

File: aiger/aig.py [hunk 3]
```diff
@@ -0,0 +1,250 @@
...
+
+
+def inverter(inputs, outputs=None):
+    if outputs is None:
+        outputs = inputs
+    else:
+        assert len(outputs) == len(inputs)
+
+    return AIG(
+        inputs=tuple(inputs),
```
Comment (MarkusRabe) [inline aiger/aig.py]: frozenset

**Comment by mvcisback** [comment]
@MarkusRabe Ready to be merged. Please review.

File: aiger/parser.py [hunk 4]
```diff
@@ -1,9 +1,13 @@
+from collections import namedtuple
```
Comment (MarkusRabe) [inline aiger/parser.py]: Not sure how Visitor class exactly interacts with the grammar. Are Symbol and SymbolTable only local? If so, they should get an underscore.

File: aiger/aig.py [hunk 5]
```diff
@@ -0,0 +1,396 @@
+from collections import defaultdict
+from itertools import chain
+from functools import reduce
+from typing import Tuple, FrozenSet, NamedTuple, Union, Mapping, List
+
+import funcy as fn
+import lenses.hooks  # TODO: remove on next lenses version release.
+from lenses import bind, lens
+from toposort import toposort
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: For `__call__` it's probably not necessary. For `_to_aig` it really simplifies the implementation.

File: aiger/aig.py [hunk 6]
```diff
@@ -0,0 +1,396 @@
...
+    return frozenset(iterable)
+
+
+class AndGate(NamedTuple):
+    left: 'Node'  # TODO: replace with Node once 3.7 lands.
+    right: 'Node'
+
+    @property
+    def children(self):
+        return tuple((self.left, self.right))
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Oh, this was copy paste being frozen set. Will fix.

File: aiger/aig.py [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+    @property
+    def outputs(self):
+        return frozenset(fn.pluck(0, self.top_level))
+
+    @property
+    def latches(self):
+        return frozenset(bind(self).Recur(Latch).collect())
+
+    @property
+    def cones(self):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Yields all cones of influence. Perhaps it should be private. Will open issue to explore.

File: aiger/aig.py [hunk 8]
```diff
@@ -0,0 +1,396 @@
...
+        return frozenset(fn.pluck(1, self.top_level))
+
+    def __rshift__(self, other):
+        return seq_compose(self, other)
+
+    def __or__(self, other):
+        return par_compose(self, other)
+
+    @property
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: I went with a property to just indicate that it was inherent to the AIGER circuit and didn't depend on state or input. I'll think about this.

File: aiger/aig.py [hunk 9]
```diff
@@ -0,0 +1,396 @@
...
+    def _eval_order(self):
+        return list(toposort(_dependency_graph(self.cones)))
+
+    def __call__(self, inputs, latches=None):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: That's a good question. I'll think about it during the performance evaluation.

File: aiger/aig.py [hunk 10]
```diff
@@ -0,0 +1,396 @@
...
+    def unroll(self, horizon, *, init=True, omit_latches=True):
+        # TODO:
+        # - Check for name collisions.
+        latches = self.latches
+        aag0 = self.cutlatches({l.name for l in latches})
+
+        def _unroll():
+            prev = aag0
+            for t in range(1, horizon + 1):
+                tmp = prev['i', {k: f"{k}##time_{t-1}" for k in aag0.inputs}]
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Sure.

File: aiger/aig.py [hunk 11]
```diff
@@ -0,0 +1,396 @@
...
+        aag, _, l_map = _to_aag(self.cones, AAG({}, {}, {}, [], self.comments))
+        aag.outputs.update({k: l_map[cone] for k, cone in self.top_level})
+        return aag
+
+    def write(self, path):
+        with open(path, 'w') as f:
+            f.write(repr(self))
+
+
+def to_idx(lit):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Good call.

File: aiger/aig.py [hunk 12]
```diff
@@ -0,0 +1,396 @@
...
+    return lit >> 1
+
+
+def _polarity(i):
+    return Inverter if inverted(i) else lambda x: x
+
+
+def inverted(lit):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Good call.

File: aiger/aig.py [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+    latches: Mapping[str, Tuple[int]]
+    outputs: Mapping[str, int]
+    gates: List[Tuple[int]]
+    comments: Tuple[str]
+
+    @property
+    def header(self):
+        max_idx = max(
+            map(to_idx,
+                chain(self.inputs.values(),
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Good catch. I guess in all of my tests, the output was always the largest literal so it still worked....

File: aiger/aig.py [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+            out += '\n'.join(f"l{idx} {name}"
+                             for idx, name in enumerate(latch_names)) + '\n'
+
+        if self.comments:
+            out += 'c\n' + '\n'.join(self.comments)
+            if out[-1] != '\n':
+                out += '\n'
+        return out
+
+    def _to_aig(self):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Goes from the literal format to the new cone of influence format. It's mostly used by the parser, but it's also useful for tests.

File: aiger/aig.py [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+             for _, (a, b, _) in self.latches.items()})
+
+        lookup = {v[0] & -2: ('AND', v) for v in self.gates}
+        lookup.update(
+            {v[0] & -2: ('LATCH', (v, k))
+             for k, v in self.latches.items()})
+        return list(toposort(deps)), lookup
+
+
+def _to_aag(gates, aag: AAG = None, *, max_idx=1, lit_map=None):
```
[the remark this answers is not in the document]
Comment (mvcisback) [reply aiger/aig.py]: Converts to the object representing the ascii encoding. It outputs the cone of influence encoding `AIG`.

It's used for pretty printing (see `__repr__`). It's also used in testing.

File: aiger/parser.py (hunk 4 above)
Comment (mvcisback) [reply aiger/parser.py]: Yeah, they come a old version of the AIG representation, but ended up being nice for bookkeeping. I'll make them private.

**Pull request merged**